### PR TITLE
fix(ilp): fix NPE in ILP ingestion during shutdown

### PR DESCRIPTION
**Pull request by bziobrowski**
Fixes NPE that happened while ingestion was doing on while qdb cloud instance was shutting down . 

I couldn't reproduce the issue so test is pretty hacky . 

Cloud logs :

```
2022-11-09 11:24:02	2022-11-09T10:24:02.383452Z I server-main cleaned worker [name=shared, worker=2, total=6]
2022-11-09 11:24:02	2022-11-09T10:24:02.383369Z I i.q.c.h.p.QueryCache closed
2022-11-09 11:24:02	2022-11-09T10:24:02.383338Z I i.q.c.h.p.QueryCache closed
2022-11-09 11:24:02	2022-11-09T10:24:02.334325Z I server-main os scheduled worker stopped [name=questdb-shared-3]
2022-11-09 11:24:02	2022-11-09T10:24:02.334320Z I server-main cleaned worker [name=shared, worker=1, total=6]
2022-11-09 11:24:02	2022-11-09T10:24:02.334228Z I i.q.c.h.p.QueryCache closed
2022-11-09 11:24:02	2022-11-09T10:24:02.334205Z I i.q.c.h.p.QueryCache closed
2022-11-09 11:24:02	2022-11-09T10:24:02.334051Z I tcp-line-server scheduling disconnect [fd=22, reason=0]
2022-11-09 11:24:02	2022-11-09T10:24:02.333660Z I i.q.WorkerPoolManager closing shared pool [name=shared, workers=6]
2022-11-09 11:24:02	2022-11-09T10:24:02.333638Z I i.q.c.p.WriterPool << [table=`ilp_test`, thread=14]
2022-11-09 11:24:02	]
2022-11-09 11:24:02		at io.questdb.mp.Worker.run(Worker.java:118)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpNetworkIOJob.run(LineTcpNetworkIOJob.java:98)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpNetworkIOJob.handleIO(LineTcpNetworkIOJob.java:123)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpAuthConnectionContext.handleIO(LineTcpAuthConnectionContext.java:87)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpConnectionContext.handleIO(LineTcpConnectionContext.java:139)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpConnectionContext.parseMeasurements(LineTcpConnectionContext.java:149)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler.scheduleEvent(LineTcpMeasurementScheduler.java:356)
2022-11-09 11:24:02		at io.questdb.cutlass.line.tcp.LineTcpMeasurementEvent.createMeasurementEvent(LineTcpMeasurementEvent.java:254)
2022-11-09 11:24:02	java.lang.NullPointerException: Cannot invoke "io.questdb.cutlass.line.tcp.TableUpdateDetails$ThreadLocalDetails.resetProcessedColumnsTracking()" because "localDetails" is null
2022-11-09 11:24:02	2022-11-09T10:24:02.333590Z C i.q.c.l.t.LineTcpConnectionContext [22] could not process line data [table=ilp_test, ex=
2022-11-09 11:24:02	2022-11-09T10:24:02.333548Z I i.q.c.l.t.TableUpdateDetails closing table parsers [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333508Z I i.q.c.l.t.TableUpdateDetails closing table parsers [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333463Z I i.q.c.l.t.TableUpdateDetails closing table parsers [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333420Z I i.q.c.l.t.TableUpdateDetails closing table parsers [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333370Z I i.q.c.l.t.TableUpdateDetails closing table parsers [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333263Z I i.q.c.l.t.TableUpdateDetails closing table parsers [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333259Z I i.q.c.l.t.TableUpdateDetails closing table writer [tableName=ilp_test]
2022-11-09 11:24:02	2022-11-09T10:24:02.333179Z I i.q.c.l.t.LineTcpWriterJob line protocol writer closing [threadId=0]
2022-11-09 11:24:02	2022-11-09T10:24:02.333140Z I server-main os scheduled worker stopped [name=questdb-ilpwriter-8]
2022-11-09 11:24:02	2022-11-09T10:24:02.333075Z I server-main cleaned worker [name=ilpwriter, worker=0, total=1]
2022-11-09 11:24:02	2022-11-09T10:24:02.288287Z I i.q.WorkerPoolManager closing dedicated pool [name=ilpwriter, workers=1]
2022-11-09 11:24:02	Pre-touch magic number: 0
2022-11-09 11:24:02	QuestDB is shutting down...
```

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 13 / 13 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiver.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|2|2|100.00%|

**Comment by bluestreak01** [comment]
@bziobrowski it would help review the PR if I could better understand the problem. May be inclusion on a log, stack trace in PR description would provide better context

File: core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpReceiver.java [hunk 1]
```diff
@@ -40,6 +40,7 @@
 
 
 public class LineTcpReceiver implements Closeable {
+    private static final SchedulerInitalizer DEFAULT_INITIALIZER = LineTcpMeasurementScheduler::new;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpReceiver.java:43]: do we defer constructor for testing only?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpReceiver.java [hunk 2]
```diff
@@ -93,6 +104,16 @@ void setSchedulerListener(SchedulerListener listener) {
         scheduler.setListener(listener);
     }
 
+    @FunctionalInterface
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpReceiver.java:107]: the warnings can be fixed with method and interface visibility

File: core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java [hunk 3]
```diff
@@ -306,6 +306,9 @@ void createMeasurementEvent(
     ) {
         writerWorkerId = LineTcpMeasurementEventType.ALL_WRITERS_INCOMPLETE_EVENT;
         final TableUpdateDetails.ThreadLocalDetails localDetails = tableUpdateDetails.getThreadLocalDetails(workerId);
+        if (localDetails == null || tableUpdateDetails.isClosed()) {
+            throw CairoException.nonCritical().put("ILP listener is closing!");
```
[the remark this answers is not in the document]
Comment (glasstiger) [reply core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java:310]: if we cannot guarantee that IO jobs are stopped before writer jobs we could just drop the event, maybe print a warning to the log that we received something during shutdown and ignoring it.
however, i think we should be able to shut ILP down orderly.

[the remark this answers is not in the document]
Comment (glasstiger) [reply core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java:310]: maybe the shared pool was used and that mixed the ordering up?

[the remark this answers is not in the document]
Comment (glasstiger) [reply core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java:310]: based on the attached logs i think the problem was that IO jobs did not have their dedicated pool and writer jobs were stopped before io jobs.
i guess we should document that if writer jobs have their own dedicated pool IO jobs should have their own too.

**Comment by bziobrowski** [comment]
I reworked the PR to prevent dedicated & shared pool mixup in ilp config .

File: core/src/main/java/io/questdb/PropServerConfiguration.java [hunk 4]
```diff
@@ -964,6 +965,14 @@ public PropServerConfiguration(
                     ilpTcpWorkerCount = 6;
                 }
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/PropServerConfiguration.java:966]: Assuming enough cpus , default for io pool is higher than for writers so maybe prioritising number of dedicated io workers would make more sense ?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpReceiver.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 9 / 9 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|1|1|100.00%|

**Review by glasstiger — APPROVED** [review]

**Pull request merged**